File: script/data_in.py
# Copyright 2025 Maktab-e-Digital Systems Lahore.
# Licensed under the Apache License, Version 2.0, see LICENSE file for details.
# SPDX-License-Identifier: Apache-2.0
#
# Description:
# This Python script converts an RGB image (specifically resized to 96×96 pixels) into a SystemVerilog-compatible stimulus file
# by extracting the red, green, and blue values of each pixel and formatting them as 24-bit binary values, where blue occupies bits [23:16], green [15:8], 
# and red [7:0]. Each line in the output file assigns a pixel to the data_in signal in the format data_in <= 24'b...; followed by a delay (#10000;),
# making it suitable for use in a Verilog or SystemVerilog testbench for simulating an image-processing module like a JPEG encoder.
#
# Author:Navaal Noshi
# Date:30th July,2025

from PIL import Image
import numpy as np

# --- Configuration ---
image_path = "desert.jpg"   # input image
output_file = "pixel_data.txt"
block_size = 8

# --- Read the image (like MATLAB imread) ---
img = np.array(Image.open(image_path))   # RGB

# --- Convert to BGR (same as MATLAB img(:,:, [3 2 1])) ---
imgBGR = img[:, :, [2, 1, 0]]

# Image size
h, w, _ = imgBGR.shape
total_pixels = h * w
pixel_count = 0

with open(output_file, "w") as f:
    # Go through blocks
    for by in range(0, h, block_size):
        for bx in range(0, w, block_size):
            # Extract one 8x8 block (same as MATLAB slice)
            block = imgBGR[by:by+block_size, bx:bx+block_size, :]

            # If this is the very last block → assert EOF once
            if pixel_count == total_pixels - 64:
                f.write("end_of_file_signal  <= 1'b1;\n")

            # Loop through 8x8 block pixels, row by row (left→right, top→down)
            for row in range(block_size):
                for col in range(block_size):
                    B, G, R = block[row, col]   # direct pixel read
                    pixel_count += 1

                    # Convert each channel to 8-bit binary
                    Rbin = format(R, "08b")
                    Gbin = format(G, "08b")
                    Bbin = format(B, "08b")

                    # BGR order
                    binStr = f"{Bbin}{Gbin}{Rbin}"

                    # Write pixel to file
                    f.write(f"\tdata_in <= 24'b{binStr};\n#10000;\n")

            # Only add block-separator if NOT the last block
            if pixel_count < total_pixels:
                f.write("#130000;\n")
                f.write("enable <= 1'b0;\n")
                f.write("#10000;\n")
                f.write("enable <= 1'b1;\n")

    # At the extreme end, only these three lines
    f.write("#130000;\n")
    f.write("enable <= 1'b0;\n")

print("✅ Pixel data written to pixel_data.txt (8x8 blocks, row-major, EOF only before last block)")

File: script/data_in.m
% Read the image
img = imread('desert.jpg');   % or .jpg, .png, etc.

% Convert to BGR (since your Verilog seems to need it as BGR)
imgBGR = img(:, :, [3 2 1]);

% Image size
[h, w, ~] = size(imgBGR);
block_size = 8;  % 8x8 pixels per block

% Open a file to save Verilog-style statements
fid = fopen('pixel_data.txt', 'w');

% Total number of pixels
total_pixels = h * w;
pixel_count = 0;

% Go through blocks
for by = 1:block_size:h
    for bx = 1:block_size:w
        % Extract one 8x8 block
        block = imgBGR(by:by+block_size-1, bx:bx+block_size-1, :);

        % Flatten block in row-major order (left→right, top→down inside block)
        pixels = reshape(permute(block, [2 1 3]), [], 3);

        % If this is the very last block → assert EOF signal once
        if pixel_count == total_pixels - 64
            fprintf(fid, 'end_of_file_signal  <= 1''b1;\n');
        end

        % Loop through 64 pixels of this block
        for k = 1:size(pixels,1)
            pixel_count = pixel_count + 1;

            B = pixels(k,1);
            G = pixels(k,2);
            R = pixels(k,3);

            % Convert each channel to 8-bit binary
            Rbin = dec2bin(R, 8);
            Gbin = dec2bin(G, 8);
            Bbin = dec2bin(B, 8);

            % Choose format (BGR)
            binStr = [Bbin Gbin Rbin];   % BGR order

            % Write pixel to file
            fprintf(fid, '\tdata_in <= 24''b%s;\n#10000;\n', binStr);
        end

        % Only add block-separator if NOT the last block
        if pixel_count < total_pixels
            fprintf(fid, '#130000;\n');
            fprintf(fid, 'enable <= 1''b0;\n');
            fprintf(fid, '#10000;\n');
            fprintf(fid, 'enable <= 1''b1;\n');
        end
    end
end

% At the extreme end, only these three lines
fprintf(fid, '#130000;\n');
fprintf(fid, 'enable <= 1''b0;\n');

fclose(fid);
disp('✅ Pixel data written to pixel_data.txt (EOF only before last block, clean end)');
